libraries/AP_NavEKF3/AP_NavEKF3_Measurements.cpp

```diff
@@ -1024,6 +1027,104 @@ void NavEKF3_core::writeExtNavVelData(const Vector3f &vel, float err, uint32_t t
     storedExtNavVel.push(extNavVelNew);
 }
 
+/*
+  update the GPS selection
+ */
+void NavEKF3_core::update_gps_selection(void)
+{
+    auto &gps = AP::gps();
+
+    // in normal operation use the primary GPS
+    selected_gps = gps.primary_sensor();
+    preferred_gps = selected_gps;
+
+    if (frontend->_affinity & EKF_AFFINITY_GPS) {
+        if (core_index < gps.num_sensors() ) {
+            // always prefer our core_index, unless we don't have that
+            // many GPS sensors available
+            preferred_gps = core_index;
+        }
+        if (gps.status(preferred_gps) >= AP_GPS::GPS_OK_FIX_3D) {
+            // select our preferred_gps if it has a 3D fix, otherwise
+            // use the primary GPS
+            selected_gps = preferred_gps;
+        }
+    }
+}
+
+/*
+  update the mag selection
+ */
+void NavEKF3_core::update_mag_selection(void)
+{
+    const auto *compass = _ahrs->get_compass();
+    if (compass == nullptr) {
+        return;
+    }
+
+    if (frontend->_affinity & EKF_AFFINITY_MAG) {
+        if (core_index < compass->get_count() &&
+            compass->healthy(core_index) &&
+            compass->use_for_yaw(core_index)) {
+            // use core_index compass if it is healthy
+            magSelectIndex = core_index;
+        }
+    }
+}
+
+/*
+  update the baro selection
+ */
+void NavEKF3_core::update_baro_selection(void)
+{
+    auto &baro = AP::baro();
+
+    // in normal operation use the primary baro
+    selected_baro = baro.get_primary();
+
+    if (frontend->_affinity & EKF_AFFINITY_BARO) {
+        if (core_index < baro.num_instances() &&
+            baro.healthy(core_index)) {
+            // use core_index baro if it is healthy
+            selected_baro = core_index;
+        }
+    }
+}
+
+/*
+  update the airspeed selection
+ */
+void NavEKF3_core::update_airspeed_selection(void)
+{
+    const auto *arsp = AP::airspeed();
+    if (arsp == nullptr) {
+        return;
+    }
+
+    // in normal operation use the primary airspeed sensor
+    selected_airspeed = arsp->get_primary();
+
+    if (frontend->_affinity & EKF_AFFINITY_ARSP) {
+        if (core_index < arsp->get_num_sensors() &&
+            arsp->healthy(core_index) &&
+            arsp->use(core_index)) {
+            // use core_index airspeed if it is healthy
+            selected_airspeed = core_index;
+        }
+    }
```

Comment:
suggest moving the non affinity case as an else condition here to avoid unnecessary airspeed object access when doing airspeed affinity, eg:

    if (frontend->_affinity & EKF_AFFINITY_ARSP &&
        core_index < arsp->get_num_sensors() &&
        arsp->healthy(core_index) &&
        arsp->use(core_index))
    {
        // use core_index airspeed if it is healthy
        selected_airspeed = core_index;
    } else {
        // in normal operation use the primary airspeed sensor
        selected_airspeed = arsp->get_primary();
    }